### PR TITLE
Enforce LinkML validation in SSSOM

### PR DESCRIPTION
**Pull request by hrshdhgd**
- [x] Edit test files to conform to linkml schema (`sssom-schema`)
- [x] Make sure validation passes.
- [x] Write corresponding tests.

### CODE REVIEW

**Comment by hrshdhgd** [comment]
I've tagged you in the commit @matentzn , in spite of converting the `msdf.df["object_match_field"]` into a list, the validation is complaining

File: sssom/constants.py [hunk 1]
```diff
@@ -47,6 +47,11 @@
 MULTIVALUED_SLOTS = [
     c for c in SCHEMA_VIEW.all_slots() if SCHEMA_VIEW.get_slot(c).multivalued
 ]
+ENTITY_REFERENCE_SLOTS = [
+    c
+    for c in SCHEMA_VIEW.all_slots()
+    if SCHEMA_VIEW.get_slot(c).range == "EntityReference"
```
Comment (matentzn) [inline sssom/constants.py:53]: us a token here again, ENTITYREFERENCE in constants.py

File: sssom/validators.py [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+from sssom_schema import MappingSet
+
+from sssom.parsers import to_mapping_set_document
+from sssom.util import MappingSetDataFrame, get_prefix_from_curie
+
+from .constants import ENTITY_REFERENCE_SLOTS, SCHEMA_VIEW, SCHEMA_YAML
+
+
+def json_schema_validate(msdf: MappingSetDataFrame) -> bool:
+    """_summary_
```
Comment (matentzn) [inline sssom/validators.py:14]: add nice summary here

File: sssom/io.py [hunk 3]
```diff
@@ -107,12 +114,11 @@ def validate_file(input_path: str) -> bool:
     :param input_path: The path to the input file in one of the legal formats, eg obographs, aligmentapi-xml
     :returns: True if valid SSSOM, false otherwise.
     """
-    try:
```
Comment (matentzn) [inline sssom/io.py:110]: extend validate_file() method to parameter to 

```
validate_file(input_path: str, validation_types = DEFAULT_VALIDATION_TYPES: list[str]) -> bool
```

in constants.py (or wherever our config resides):

```
class SchemaValidationType(Enum):
    JsonSchema = 1
    Shacl = 2
    PrefixMapCompleteness = 3

DEFAULT_VALIDATION_TYPES = [ SchemaValidationType.JsonSchema, SchemaValidationType.PrefixMapCompleteness ]
```

File: sssom/io.py [hunk 4]
```diff
@@ -107,12 +114,11 @@ def validate_file(input_path: str) -> bool:
...
-        parse_sssom_table(file_path=input_path)
-        return True
-    except Exception as e:
-        logging.exception("The file is invalid", e)
-        return False
+    # Two things to check:
+    # 1. All prefixes in the DataFrame are define in prefix_map
+    # 2. All columns in the DataFrame abide by sssom-schema.
+    msdf = parse_sssom_table(file_path=input_path)
+    json_schema_validate(msdf)
```
Comment (matentzn) [inline sssom/io.py:121]: rename to validate() and pass in msdf and `validation_types`.

File: sssom/validators.py [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+from linkml.validators.jsonschemavalidator import JsonSchemaDataValidator
+from sssom_schema import MappingSet
+
+from sssom.parsers import to_mapping_set_document
+from sssom.util import MappingSetDataFrame, get_prefix_from_curie
+
+from .constants import ENTITY_REFERENCE_SLOTS, SCHEMA_VIEW, SCHEMA_YAML
+
+
+def json_schema_validate(msdf: MappingSetDataFrame) -> bool:
```
Comment (matentzn) [inline sssom/validators.py:13]: keep the json_schema_validate method but also add a generic

```
validate(input_path: str, validation_types = DEFAULT_VALIDATION_TYPES: list[str]) -> bool
```

In this method, add a switch over all the `validation_types`, and when say case `SchemaValidationType.JsonSchema` call the schema_validation method.

File: sssom/io.py [hunk 6]
```diff
@@ -107,12 +114,11 @@ def validate_file(input_path: str) -> bool:
     :param input_path: The path to the input file in one of the legal formats, eg obographs, aligmentapi-xml
     :returns: True if valid SSSOM, false otherwise.
     """
-    try:
-        parse_sssom_table(file_path=input_path)
-        return True
-    except Exception as e:
-        logging.exception("The file is invalid", e)
-        return False
```
Comment (matentzn) [inline sssom/io.py:115]: remove the --> Bool  return type from validate_file(). Only `validate()` should have a return type boolean, validate_file just calls `validate()` as you already do here..

File: sssom/validators.py [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+    metadata_keys = list(msdf.metadata.keys())
+    df_columns_list = msdf.df.columns.to_list()
+    all_keys = metadata_keys + df_columns_list
+    ent_ref_slots = [s for s in all_keys if s in ENTITY_REFERENCE_SLOTS]
+    prefix_list = []
+    for slot in ent_ref_slots:
+        if slot in metadata_keys:
+            if type(msdf.metadata[slot]) == list:
+                for s in msdf.metadata[slot]:
+                    if get_prefix_from_curie(s) == "":
```
Comment (matentzn) [inline sssom/validators.py:63]: This is ok here (keep it as you have it), but we should make sure that json_schema_validation catches this case!

File: tests/test_validate.py [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+
+
+class TestValidate(unittest.TestCase):
+    """A test case for sorting msdf columns."""
+
+    def setUp(self) -> None:
+        """Test up the test cases with the third basic example."""
+        self.msdf = parse_sssom_table(f"{data_dir}/basic.tsv")
+
+    def test_validate(self):
```
Comment (matentzn) [inline tests/test_validate.py:19]: Lets try and add 4 good tests here.. I think this command may be the most important of all in the end :)

**Review by matentzn — CHANGES_REQUESTED** [review]
Excellent @hrshdhgd very nice! Few suggestions here and I think this is good to go!

File: sssom/validators.py (hunk 2 above)
Comment (hrshdhgd) [reply sssom/validators.py:14]: Done.

File: sssom/validators.py (hunk 7 above)
Comment (hrshdhgd) [reply sssom/validators.py:63]: This was just for debugging purposes. I will `raise Exception` to catch it.

File: sssom/cli.py [hunk 9]
```diff
@@ -178,9 +179,13 @@ def parse(
 
 @main.command()
 @input_argument
-def validate(input: str):
+@click.option(
+    "--validation-types", "-V", type=click.Choice(SchemaValidationType), multiple=True
```
Comment (matentzn) [inline sssom/cli.py:183]: Add default _all_ schema checks

File: sssom/parsers.py [hunk 10]
```diff
@@ -311,6 +311,10 @@ def _get_mdict_ms_and_bad_attrs(
     mdict = {}
 
     for k, v in row.items():
+        if k == "creator_id":
+            import pdb
```
[the remark this answers is not in the document]
Comment (matentzn) [reply sssom/parsers.py:315]: Is this still there?

File: tests/test_reconcile.py [hunk 11]
```diff
@@ -31,7 +31,7 @@ def test_merge(self):
         msdf2 = parse_sssom_table(data_dir / "basic2.tsv")
         merged_msdf = merge_msdf(msdf1, msdf2)
 
-        self.assertEqual(71, len(merged_msdf.df))
+        self.assertEqual(122, len(merged_msdf.df))
```
Comment (matentzn) [inline tests/test_reconcile.py:34]: Seems odd they doubled in size?

File: tests/test_validate.py [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+
+    def setUp(self) -> None:
+        """Test up the test cases with the third basic example."""
+        self.correct_msdf1 = parse_sssom_table(f"{data_dir}/basic.tsv")
+        self.bad_msdf1 = parse_sssom_table(f"{data_dir}/bad_basic.tsv")
+        self.validation_types = DEFAULT_VALIDATION_TYPES
+        self.shacl_validation_types = [SchemaValidationType.Shacl]
+
+    def test_validate_json(self):
+        """Test JSONSchemaValidation."""
```
Comment (matentzn) [inline tests/test_validate.py:24]: make a better comment saying why this should fail or pass.

File: tests/test_validate.py [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+        self.assertIsNone(validate(self.correct_msdf1, self.validation_types))
+
+    def test_validate_json_fail(self):
+        """Test if JSONSchemaValidation fail is as expected."""
```
Comment (matentzn) [inline tests/test_validate.py:28]: make a better comment saying why this should fail or pass.

File: tests/test_validate.py [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+        self.assertRaises(
+            ValidationError, validate, self.bad_msdf1, self.validation_types
+        )
+
+    def test_validate_shacl(self):
+        """Test Shacl validation (Not implemented)."""
```
Comment (matentzn) [inline tests/test_validate.py:34]: make a better comment saying why this should fail or pass.

File: tox.ini [hunk 15]
```diff
@@ -12,10 +12,10 @@ envlist =
     py
 
 [testenv]
-whitelist_externals = 
-    pytest
```
Comment (matentzn) [inline tox.ini:16]: Why this change?

File: tox.ini [hunk 16]
```diff
@@ -12,10 +12,10 @@ envlist =
...
 commands =
-    pytest
+    python -m pytest
```
Comment (matentzn) [inline tox.ini:16]: Why this change?

File: tox.ini [hunk 17]
```diff
@@ -12,10 +12,10 @@ envlist =
...
+deps = 
+    linkml
```
Comment (matentzn) [inline tox.ini:18]: Why this change?

File: sssom/util.py [hunk 18]
```diff
@@ -714,13 +714,12 @@ def inject_metadata_into_df(msdf: MappingSetDataFrame) -> MappingSetDataFrame:
...
         schema = yaml.safe_load(file)
     slots = schema["classes"]["mapping"]["slots"]
-
     if msdf.metadata is not None and msdf.df is not None:
         for k, v in msdf.metadata.items():
             if k not in msdf.df.columns and k in slots:
                 if k == MAPPING_SET_ID:
                     k = MAPPING_SET_SOURCE
-                msdf.df[k] = v
+                msdf.df[k] = str(v)
```
[the remark this answers is not in the document]
Comment (hrshdhgd) [reply sssom/util.py:722]: Yes.

File: sssom/parsers.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (hrshdhgd) [reply sssom/parsers.py:315]: Aah... nice catch.Thanks!

File: sssom/cli.py (hunk 9 above)
Comment (hrshdhgd) [reply sssom/cli.py:183]: Done

File: tests/test_reconcile.py (hunk 11 above)
Comment (hrshdhgd) [reply tests/test_reconcile.py:34]: Some of the `mapping_justification` values were changed in such a way that I think that's expected.

File: tox.ini (hunk 15 above)
Comment (hrshdhgd) [reply tox.ini:16]: This enables `tox` to use the purest within the original environment rather than installing (a potentially different version) within its own environment. Avoids confusion.

File: tox.ini (hunk 16 above)
Comment (hrshdhgd) [reply tox.ini:16]: This keeps it uniform whether one runs `pytest` from the terminal or `python-m pytest` from within `tox`.

File: tests/test_validate.py (hunk 12 above)
Comment (hrshdhgd) [reply tests/test_validate.py:24]: Done.

File: tests/test_validate.py (hunk 13 above)
Comment (hrshdhgd) [reply tests/test_validate.py:28]: Done

File: sssom/parsers.py [hunk 19]
```diff
@@ -828,7 +831,7 @@ def to_mapping_set_document(msdf: MappingSetDataFrame) -> MappingSetDocument:
     if msdf.metadata is not None:
         for k, v in msdf.metadata.items():
             if k != PREFIX_MAP_KEY:
-                ms[k] = v
+                ms[k] = _address_multivalued_slot(k, str(v))
```
[the remark this answers is not in the document]
Comment (hrshdhgd) [reply sssom/parsers.py]: Yes, we did it together on the call. The change you suggested in #274 works as expected in this PR.

**Review by matentzn — APPROVED** [review]
